From db114c3c684e2a6a12a25ac89f886ae1a1786fd4 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 11 Aug 2023 14:36:05 +0200
Subject: [PATCH 0443/1695] pta: scmi: support OCall with SCP firmware

Updates SCMI PTA to support OCall context with SCP-firmware
implementation of the SCMI services.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 08d7ded4e99cfbda3aa882a655de4d608dbed537 pta: scmi: support OCall with SCP firmware]
Change-Id: I76d0b6a3c7e93a959d41fff035f916ea3108cf25
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 core/pta/scmi.c | 32 ++++++++++++++++++++------------
 1 file changed, 20 insertions(+), 12 deletions(-)

diff --git a/core/pta/scmi.c b/core/pta/scmi.c
index 300a2f3dd..a0efcc073 100644
--- a/core/pta/scmi.c
+++ b/core/pta/scmi.c
@@ -207,16 +207,29 @@ static enum optee_scmi_ocall_reply pta_scmi_ocall(uint32_t channel_id,
 		break;
 	case PTA_SCMI_OCALL_PROCESS_MSG_CHANNEL:
 		FMSG("Posting MSG message on channel %u"PRIu32, channel_id);
-		if (!IS_ENABLED(CFG_SCMI_MSG_SHM_MSG) || !dyn_shm)
+		if (!dyn_shm)
 			return PTA_SCMI_OCALL_ERROR;
 
 		dyn_shm->in_size = ocall2_params[1];
 		dyn_shm->out_size = dyn_shm->out_max_size;
 
-		if (scmi_msg_threaded_entry(channel_id, dyn_shm->in_buf,
-					    dyn_shm->in_size, dyn_shm->out_buf,
-					    &dyn_shm->out_size))
-			return PTA_SCMI_OCALL_ERROR;
+		if (IS_ENABLED(CFG_SCMI_MSG_SHM_MSG)) {
+			if (scmi_msg_threaded_entry(channel_id, dyn_shm->in_buf,
+						    dyn_shm->in_size,
+						    dyn_shm->out_buf,
+						    &dyn_shm->out_size))
+				return PTA_SCMI_OCALL_ERROR;
+		}
+
+		if (IS_ENABLED(CFG_SCMI_SCPFW)) {
+			if (scmi_server_msg_process_thread(channel_id,
+							   dyn_shm->in_buf,
+							   dyn_shm->in_size,
+							   dyn_shm->out_buf,
+							   &dyn_shm->out_size))
+				return PTA_SCMI_OCALL_ERROR;
+		}
+
 		break;
 	case PTA_SCMI_OCALL_CLOSE_THREAD:
 		FMSG("Closing channel %u"PRIu32, channel_id);
@@ -274,7 +287,6 @@ static TEE_Result cmd_scmi_ocall_msg_thread(uint32_t ptypes,
 						TEE_PARAM_TYPE_MEMREF_INPUT,
 						TEE_PARAM_TYPE_MEMREF_OUTPUT,
 						TEE_PARAM_TYPE_NONE);
-
 	struct msg_shm dyn_shm = {
 		.in_buf = params[1].memref.buffer,
 		.in_size = params[1].memref.size,
@@ -282,16 +294,12 @@ static TEE_Result cmd_scmi_ocall_msg_thread(uint32_t ptypes,
 		.out_max_size = params[2].memref.size,
 	};
 	unsigned int channel_id = params[0].value.a;
-	struct scmi_msg_channel *channel = NULL;
-
-	if (!IS_ENABLED(CFG_SCMI_MSG_SHM_MSG))
-		return TEE_ERROR_NOT_SUPPORTED;
 
 	if (ptypes != exp_pt || !dyn_shm.in_buf || !dyn_shm.out_buf)
 		return TEE_ERROR_BAD_PARAMETERS;
 
-	channel = plat_scmi_get_channel(channel_id);
-	if (!channel)
+	if (IS_ENABLED(CFG_SCMI_MSG_SHM_MSG) &&
+	    !plat_scmi_get_channel(channel_id))
 		return TEE_ERROR_BAD_PARAMETERS;
 
 	while (1) {
-- 
2.47.2

